File: src/c/tests/test_crossimpl.c
/*
 * Cross-Implementation Test Runner
 *
 * Runs test cases matching spec/tests/move_validation.json to verify
 * C implementation matches the shared spec.
 *
 * NOTE: Since C doesn't have a built-in JSON parser, these tests are
 * manually coded to match the spec file. Keep in sync with:
 * spec/tests/move_validation.json
 *
 * Signed-by: agent #29 claude-sonnet-4 via opencode 20260122T08:15:15
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#include "../board.h"
#include "../moves.h"

/* Test counters */
static int tests_run = 0;
static int tests_passed = 0;

#define TEST(name) static void test_##name(void)
#define RUN_TEST(name) do { \
    printf("  Running %s...", #name); \
    test_##name(); \
    tests_run++; \
    tests_passed++; \
    printf(" PASSED\n"); \
} while(0)

#define ASSERT(cond) do { \
    if (!(cond)) { \
        printf(" FAILED at line %d: %s\n", __LINE__, #cond); \
        exit(1); \
    } \
} while(0)

#define ASSERT_EQ(a, b) do { \
    if ((a) != (b)) { \
        printf(" FAILED at line %d: %d != %d\n", __LINE__, (int)(a), (int)(b)); \
        exit(1); \
    } \
} while(0)

/* ============================================================================
 * Helper functions
 * ============================================================================ */

/* Helper to check if a move is legal */
static bool check_move_legal(Board* board, Cell from, Cell to) {
    Move m = {from, to, PIECE_NONE};
    return is_move_legal(board, m);
}

/* Helper to check if a move captures */
static bool check_move_captures(Board* board, Cell from, Cell to) {
    Move m = {from, to, PIECE_NONE};
    if (!is_move_legal(board, m)) return false;
    Piece* target = board_get(board, to);
    return target->type != PIECE_NONE;
}

/* ============================================================================
 * Board Validation Tests (from spec/tests/move_validation.json)
 * ============================================================================ */

/* board_001: Center cell is valid */
TEST(board_001) {
    ASSERT(cell_is_valid(cell_make(0, 0)));
}

/* board_002: Corner cell at max radius is valid */
TEST(board_002) {
    ASSERT(cell_is_valid(cell_make(4, 0)));
}

/* board_003: Cell outside board is invalid */
TEST(board_003) {
    ASSERT(!cell_is_valid(cell_make(5, 0)));
}

/* board_004: Cell violating q+r constraint is invalid */
TEST(board_004) {
    ASSERT(!cell_is_valid(cell_make(3, 3)));
}

/* board_005: Negative coordinate valid cell */
TEST(board_005) {
    ASSERT(cell_is_valid(cell_make(-4, 0)));
}

/* board_006: Corner at min q, max r */
TEST(board_006) {
    ASSERT(cell_is_valid(cell_make(-4, 4)));
}

/* board_007: Cell at q+r=-4 boundary is valid */
TEST(board_007) {
    ASSERT(cell_is_valid(cell_make(0, -4)));
}

/* board_008: Cell beyond -4 boundary is invalid */
TEST(board_008) {
    ASSERT(!cell_is_valid(cell_make(0, -5)));
}

/* ============================================================================
 * Move Validation Tests (from spec/tests/move_validation.json)
 * ============================================================================ */

/* king_001: King can move to adjacent empty cell */
TEST(king_001) {
    Board board;
    board_clear(&board);
    
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wk);
    board.white_king = cell_make(0, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 0), cell_make(1, 0)));
}

/* king_002: King cannot move 2 squares */
TEST(king_002) {
    Board board;
    board_clear(&board);
    
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wk);
    board.white_king = cell_make(0, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(!check_move_legal(&board, cell_make(0, 0), cell_make(2, 0)));
}

/* king_003: King can capture enemy piece */
TEST(king_003) {
    Board board;
    board_clear(&board);
    
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    Piece bp = {PIECE_PAWN, COLOR_BLACK, 0};
    board_set(&board, cell_make(0, 0), wk);
    board_set(&board, cell_make(1, 0), bp);
    board.white_king = cell_make(0, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 0), cell_make(1, 0)));
    ASSERT(check_move_captures(&board, cell_make(0, 0), cell_make(1, 0)));
}

/* king_004: King cannot capture friendly piece */
TEST(king_004) {
    Board board;
    board_clear(&board);
    
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    Piece wp = {PIECE_PAWN, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wk);
    board_set(&board, cell_make(1, 0), wp);
    board.white_king = cell_make(0, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(!check_move_legal(&board, cell_make(0, 0), cell_make(1, 0)));
}

/* queen_001: Queen can slide multiple squares */
TEST(queen_001) {
    Board board;
    board_clear(&board);
    
    Piece wq = {PIECE_QUEEN, COLOR_WHITE, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wq);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 0), cell_make(0, -3)));
}

/* queen_002: Queen cannot jump over pieces */
TEST(queen_002) {
    Board board;
    board_clear(&board);
    
    Piece wq = {PIECE_QUEEN, COLOR_WHITE, 0};
    Piece wp = {PIECE_PAWN, COLOR_WHITE, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wq);
    board_set(&board, cell_make(0, -1), wp);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(!check_move_legal(&board, cell_make(0, 0), cell_make(0, -3)));
}

/* pawn_001: White pawn moves north */
TEST(pawn_001) {
    Board board;
    board_clear(&board);
    
    Piece wp = {PIECE_PAWN, COLOR_WHITE, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 2), wp);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 2), cell_make(0, 1)));
}

/* pawn_002: White pawn cannot move south */
TEST(pawn_002) {
    Board board;
    board_clear(&board);
    
    Piece wp = {PIECE_PAWN, COLOR_WHITE, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 2), wp);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(!check_move_legal(&board, cell_make(0, 2), cell_make(0, 3)));
}

/* pawn_003: White pawn captures forward (north) */
TEST(pawn_003) {
    Board board;
    board_clear(&board);
    
    Piece wp = {PIECE_PAWN, COLOR_WHITE, 0};
    Piece bp = {PIECE_PAWN, COLOR_BLACK, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 2), wp);
    board_set(&board, cell_make(0, 1), bp);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 2), cell_make(0, 1)));
    ASSERT(check_move_captures(&board, cell_make(0, 2), cell_make(0, 1)));
}

/* pawn_004: White pawn captures diagonally NE */
TEST(pawn_004) {
    Board board;
    board_clear(&board);
    
    Piece wp = {PIECE_PAWN, COLOR_WHITE, 0};
    Piece bp = {PIECE_PAWN, COLOR_BLACK, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 2), wp);
    board_set(&board, cell_make(1, 1), bp);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 2), cell_make(1, 1)));
    ASSERT(check_move_captures(&board, cell_make(0, 2), cell_make(1, 1)));
}

/* pawn_005: Black pawn moves south */
TEST(pawn_005) {
    Board board;
    board_clear(&board);
    
    Piece bp = {PIECE_PAWN, COLOR_BLACK, 0};
    Piece bk = {PIECE_KING, COLOR_BLACK, 0};
    board_set(&board, cell_make(0, -2), bp);
    board_set(&board, cell_make(-4, 0), bk);
    board.black_king = cell_make(-4, 0);
    board.to_move = COLOR_BLACK;
    
    ASSERT(check_move_legal(&board, cell_make(0, -2), cell_make(0, -1)));
}

/* pawn_006: Pawn blocked by piece */
TEST(pawn_006) {
    Board board;
    board_clear(&board);
    
    Piece wp = {PIECE_PAWN, COLOR_WHITE, 0};
    Piece bp = {PIECE_PAWN, COLOR_BLACK, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 2), wp);
    board_set(&board, cell_make(0, 1), bp);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    /* Pawn can't move forward to occupied square without capture */
    /* Actually in Underchex pawns CAN capture forward, so this should be legal */
    /* Let me check: spec says pawn captures N/NE/NW. So forward capture is legal. */
    /* Test should be: pawn blocked by FRIENDLY piece */
}

/* pawn_007: Pawn blocked by friendly piece (cannot capture) */
TEST(pawn_007) {
    Board board;
    board_clear(&board);
    
    Piece wp1 = {PIECE_PAWN, COLOR_WHITE, 0};
    Piece wp2 = {PIECE_PAWN, COLOR_WHITE, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 2), wp1);
    board_set(&board, cell_make(0, 1), wp2);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(!check_move_legal(&board, cell_make(0, 2), cell_make(0, 1)));
}

/* knight_001: Knight leaps to valid target */
TEST(knight_001) {
    Board board;
    board_clear(&board);
    
    Piece wn = {PIECE_KNIGHT, COLOR_WHITE, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wn);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 0), cell_make(1, -2)));
}

/* knight_002: Knight can jump over pieces */
TEST(knight_002) {
    Board board;
    board_clear(&board);
    
    Piece wn = {PIECE_KNIGHT, COLOR_WHITE, 0};
    Piece wp = {PIECE_PAWN, COLOR_WHITE, 0};
    Piece bp = {PIECE_PAWN, COLOR_BLACK, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wn);
    board_set(&board, cell_make(0, -1), wp);
    board_set(&board, cell_make(1, -1), bp);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 0), cell_make(1, -2)));
}

/* knight_003: Knight cannot move to invalid target */
TEST(knight_003) {
    Board board;
    board_clear(&board);
    
    Piece wn = {PIECE_KNIGHT, COLOR_WHITE, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wn);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    /* Adjacent move is not a knight move */
    ASSERT(!check_move_legal(&board, cell_make(0, 0), cell_make(1, 0)));
}

/* lance_001: Lance A slides north */
TEST(lance_001) {
    Board board;
    board_clear(&board);
    
    Piece wl = {PIECE_LANCE, COLOR_WHITE, 0};  /* variant 0 = A */
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 2), wl);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 2), cell_make(0, -2)));
}

/* lance_002: Lance A cannot move NE (not in its directions) */
TEST(lance_002) {
    Board board;
    board_clear(&board);
    
    Piece wl = {PIECE_LANCE, COLOR_WHITE, 0};  /* variant 0 = A: N,S,NW,SE */
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 2), wl);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(!check_move_legal(&board, cell_make(0, 2), cell_make(2, 0)));
}

/* lance_003: Lance B slides NE */
TEST(lance_003) {
    Board board;
    board_clear(&board);
    
    Piece wl = {PIECE_LANCE, COLOR_WHITE, 1};  /* variant 1 = B: N,S,NE,SW */
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wl);
    board_set(&board, cell_make(-4, 0), wk);
    board.white_king = cell_make(-4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 0), cell_make(2, -2)));
}

/* lance_004: Lance B cannot move NW */
TEST(lance_004) {
    Board board;
    board_clear(&board);
    
    Piece wl = {PIECE_LANCE, COLOR_WHITE, 1};  /* variant 1 = B: N,S,NE,SW */
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wl);
    board_set(&board, cell_make(4, 0), wk);  /* Put king away from NW direction */
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    /* NW direction is (-1, 0) per DIRECTIONS, so (-2, 0) is 2 steps NW */
    ASSERT(!check_move_legal(&board, cell_make(0, 0), cell_make(-2, 0)));
}

/* chariot_001: Chariot slides NE */
TEST(chariot_001) {
    Board board;
    board_clear(&board);
    
    Piece wc = {PIECE_CHARIOT, COLOR_WHITE, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wc);
    board_set(&board, cell_make(-4, 0), wk);
    board.white_king = cell_make(-4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(check_move_legal(&board, cell_make(0, 0), cell_make(3, -3)));
}

/* chariot_002: Chariot cannot move north (not diagonal) */
TEST(chariot_002) {
    Board board;
    board_clear(&board);
    
    Piece wc = {PIECE_CHARIOT, COLOR_WHITE, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wc);
    board_set(&board, cell_make(-4, 0), wk);
    board.white_king = cell_make(-4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(!check_move_legal(&board, cell_make(0, 0), cell_make(0, -2)));
}

/* check_001: King cannot move into check */
TEST(check_001) {
    Board board;
    board_clear(&board);
    
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    Piece bq = {PIECE_QUEEN, COLOR_BLACK, 0};
    Piece bk = {PIECE_KING, COLOR_BLACK, 0};
    board_set(&board, cell_make(0, 0), wk);
    board_set(&board, cell_make(1, -4), bq);  /* Queen attacks (1,0) along its ray */
    board_set(&board, cell_make(-4, 0), bk);
    board.white_king = cell_make(0, 0);
    board.black_king = cell_make(-4, 0);
    board.to_move = COLOR_WHITE;
    
    /* King moving to (1,0) would be under attack from queen at (1,-4) */
    ASSERT(!check_move_legal(&board, cell_make(0, 0), cell_make(1, 0)));
}

/* turn_001: Cannot move opponent's piece */
TEST(turn_001) {
    Board board;
    board_clear(&board);
    
    Piece bp = {PIECE_PAWN, COLOR_BLACK, 0};
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, -2), bp);
    board_set(&board, cell_make(4, 0), wk);
    board.white_king = cell_make(4, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(!check_move_legal(&board, cell_make(0, -2), cell_make(0, -1)));
}

/* turn_002: Cannot move from empty cell */
TEST(turn_002) {
    Board board;
    board_clear(&board);
    
    Piece wk = {PIECE_KING, COLOR_WHITE, 0};
    board_set(&board, cell_make(0, 0), wk);
    board.white_king = cell_make(0, 0);
    board.to_move = COLOR_WHITE;
    
    ASSERT(!check_move_legal(&board, cell_make(1, 0), cell_make(2, 0)));
}

/* ============================================================================
 * Coverage Report
 * ============================================================================ */

TEST(coverage_report) {
    printf("\n\n=== Spec Test Coverage Report (C) ===\n");
    printf("Board validation tests: 8\n");
    printf("Move validation tests: 21\n");
    printf("Total spec-aligned tests: 29\n");
    printf("=====================================\n\n");
    /* This test always passes - it's just for reporting */
}

/* ============================================================================
 * Main
 * ============================================================================ */

int main(void) {
    printf("Running Underchex C cross-implementation tests...\n\n");
    
    printf("Board validation tests:\n");
    RUN_TEST(board_001);
    RUN_TEST(board_002);
    RUN_TEST(board_003);
    RUN_TEST(board_004);
    RUN_TEST(board_005);
    RUN_TEST(board_006);
    RUN_TEST(board_007);
    RUN_TEST(board_008);
    
    printf("\nMove validation tests:\n");
    RUN_TEST(king_001);
    RUN_TEST(king_002);
    RUN_TEST(king_003);
    RUN_TEST(king_004);
    RUN_TEST(queen_001);
    RUN_TEST(queen_002);
    RUN_TEST(pawn_001);
    RUN_TEST(pawn_002);
    RUN_TEST(pawn_003);
    RUN_TEST(pawn_004);
    RUN_TEST(pawn_005);
    RUN_TEST(pawn_007);
    RUN_TEST(knight_001);
    RUN_TEST(knight_002);
    RUN_TEST(knight_003);
    RUN_TEST(lance_001);
    RUN_TEST(lance_002);
    RUN_TEST(lance_003);
    RUN_TEST(lance_004);
    RUN_TEST(chariot_001);
    RUN_TEST(chariot_002);
    RUN_TEST(check_001);
    RUN_TEST(turn_001);
    RUN_TEST(turn_002);
    
    RUN_TEST(coverage_report);
    
    printf("\n========================================\n");
    printf("Cross-impl Tests: %d/%d passed\n", tests_passed, tests_run);
    printf("========================================\n");
    
    return (tests_passed == tests_run) ? 0 : 1;
}

File: src/python/tests/test_crossimpl.py
"""
Cross-Implementation Test Runner

Runs test cases from spec/tests/move_validation.json to verify
Python implementation matches the shared spec.

Signed-by: agent #28 claude-sonnet-4 via opencode 20260122T08:04:58
"""
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any, Optional

import pytest

from underchex.board import is_valid_cell
from underchex.moves import validate_move
from underchex.types import (
    HexCoord,
    Piece,
    Color,
    PieceType,
    LanceVariant,
    coord_to_string,
    BoardState,
)


# ============================================================================
# Test Suite Loading
# ============================================================================

def load_test_suite() -> dict[str, Any]:
    """Load the shared test suite from spec directory."""
    spec_path = Path(__file__).parent.parent.parent.parent / "spec" / "tests" / "move_validation.json"
    with open(spec_path, "r") as f:
        return json.load(f)


def spec_to_piece(spec: dict[str, Any]) -> Piece:
    """Convert spec piece definition to Piece object."""
    piece_type: PieceType = spec["piece"]
    color: Color = spec["color"]
    variant: Optional[LanceVariant] = spec.get("variant")
    return Piece(type=piece_type, color=color, variant=variant)


def build_board_from_spec(setup: dict[str, Any]) -> BoardState:
    """Build a board state from spec setup."""
    board: BoardState = {}
    for placement in setup["pieces"]:
        piece = spec_to_piece(placement)
        coord = HexCoord(q=placement["q"], r=placement["r"])
        board[coord_to_string(coord)] = piece
    return board


# ============================================================================
# Board Validation Tests
# ============================================================================

class TestBoardValidation:
    """Test board cell validation against spec."""
    
    @pytest.fixture(scope="class")
    def test_suite(self) -> dict[str, Any]:
        return load_test_suite()
    
    def test_center_cell_is_valid(self) -> None:
        """board_001: Center cell is valid."""
        assert is_valid_cell(HexCoord(q=0, r=0)) is True
    
    def test_corner_at_max_radius_is_valid(self) -> None:
        """board_002: Corner cell at max radius is valid."""
        assert is_valid_cell(HexCoord(q=4, r=0)) is True
    
    def test_cell_outside_board_is_invalid(self) -> None:
        """board_003: Cell outside board is invalid."""
        assert is_valid_cell(HexCoord(q=5, r=0)) is False
    
    def test_cell_violating_constraint_is_invalid(self) -> None:
        """board_004: Cell violating q+r constraint is invalid."""
        assert is_valid_cell(HexCoord(q=3, r=3)) is False
    
    def test_all_board_validation_from_spec(self, test_suite: dict[str, Any]) -> None:
        """Run all board validation tests from spec."""
        board_tests = [tc for tc in test_suite["testCases"] if tc["type"] == "boardValidation"]
        
        for tc in board_tests:
            coord = HexCoord(q=tc["input"]["q"], r=tc["input"]["r"])
            result = is_valid_cell(coord)
            expected = tc["expected"]["valid"]
            assert result == expected, f'{tc["id"]}: {tc["description"]}'


# ============================================================================
# Move Validation Tests
# ============================================================================

class TestMoveValidation:
    """Test move validation against spec."""
    
    @pytest.fixture(scope="class")
    def test_suite(self) -> dict[str, Any]:
        return load_test_suite()
    
    def test_all_move_validation_from_spec(self, test_suite: dict[str, Any]) -> None:
        """Run all move validation tests from spec."""
        move_tests = [tc for tc in test_suite["testCases"] if tc["type"] == "moveValidation"]
        
        for tc in move_tests:
            board = build_board_from_spec(tc["setup"])
            turn: Color = tc["setup"]["turn"]
            
            from_coord = HexCoord(q=tc["move"]["from"]["q"], r=tc["move"]["from"]["r"])
            to_coord = HexCoord(q=tc["move"]["to"]["q"], r=tc["move"]["to"]["r"])
            
            result = validate_move(board, from_coord, to_coord, turn)
            
            # Check legal/illegal
            assert result.legal == tc["expected"]["legal"], \
                f'{tc["id"]}: {tc["description"]} - expected legal={tc["expected"]["legal"]}, got legal={result.legal}'
            
            # If legal and capture is specified, check it
            if tc["expected"]["legal"] and "capture" in tc["expected"]:
                assert result.capture == tc["expected"]["capture"], \
                    f'{tc["id"]}: {tc["description"]} - expected capture={tc["expected"]["capture"]}, got capture={result.capture}'
            
            # If illegal with specific reason, check it
            if not tc["expected"]["legal"] and "reason" in tc["expected"]:
                assert result.reason == tc["expected"]["reason"], \
                    f'{tc["id"]}: {tc["description"]} - expected reason={tc["expected"]["reason"]}, got reason={result.reason}'


# ============================================================================
# Individual Test Cases (for better error reporting)
# ============================================================================

class TestKingMoves:
    """Test king-specific move validation."""
    
    def test_king_can_move_to_adjacent_empty_cell(self) -> None:
        """king_001: King can move to adjacent empty cell."""
        board: BoardState = {
            "0,0": Piece(type="king", color="white")
        }
        result = validate_move(board, HexCoord(0, 0), HexCoord(1, 0), "white")
        assert result.legal is True
    
    def test_king_cannot_move_two_squares(self) -> None:
        """king_002: King cannot move 2 squares."""
        board: BoardState = {
            "0,0": Piece(type="king", color="white")
        }
        result = validate_move(board, HexCoord(0, 0), HexCoord(2, 0), "white")
        assert result.legal is False
    
    def test_king_can_capture_enemy(self) -> None:
        """king_003: King can capture enemy piece."""
        board: BoardState = {
            "0,0": Piece(type="king", color="white"),
            "1,0": Piece(type="pawn", color="black")
        }
        result = validate_move(board, HexCoord(0, 0), HexCoord(1, 0), "white")
        assert result.legal is True
        assert result.capture is True


class TestQueenMoves:
    """Test queen-specific move validation."""
    
    def test_queen_can_slide_multiple_squares(self) -> None:
        """queen_001: Queen can slide multiple squares."""
        board: BoardState = {
            "0,0": Piece(type="queen", color="white")
        }
        result = validate_move(board, HexCoord(0, 0), HexCoord(0, -3), "white")
        assert result.legal is True
    
    def test_queen_cannot_jump_over_pieces(self) -> None:
        """queen_002: Queen cannot jump over pieces."""
        board: BoardState = {
            "0,0": Piece(type="queen", color="white"),
            "0,-1": Piece(type="pawn", color="white")
        }
        result = validate_move(board, HexCoord(0, 0), HexCoord(0, -3), "white")
        assert result.legal is False


class TestPawnMoves:
    """Test pawn-specific move validation."""
    
    def test_white_pawn_moves_north(self) -> None:
        """pawn_001: White pawn moves north."""
        board: BoardState = {
            "0,2": Piece(type="pawn", color="white")
        }
        result = validate_move(board, HexCoord(0, 2), HexCoord(0, 1), "white")
        assert result.legal is True
    
    def test_white_pawn_cannot_move_south(self) -> None:
        """pawn_002: White pawn cannot move south."""
        board: BoardState = {
            "0,2": Piece(type="pawn", color="white")
        }
        result = validate_move(board, HexCoord(0, 2), HexCoord(0, 3), "white")
        assert result.legal is False


class TestKnightMoves:
    """Test knight-specific move validation."""
    
    def test_knight_leaps_to_valid_target(self) -> None:
        """knight_001: Knight leaps to valid target."""
        board: BoardState = {
            "0,0": Piece(type="knight", color="white")
        }
        result = validate_move(board, HexCoord(0, 0), HexCoord(1, -2), "white")
        assert result.legal is True
    
    def test_knight_can_jump_over_pieces(self) -> None:
        """knight_002: Knight can jump over pieces."""
        board: BoardState = {
            "0,0": Piece(type="knight", color="white"),
            "0,-1": Piece(type="pawn", color="white"),
            "1,-1": Piece(type="pawn", color="black")
        }
        result = validate_move(board, HexCoord(0, 0), HexCoord(1, -2), "white")
        assert result.legal is True


class TestLanceMoves:
    """Test lance-specific move validation."""
    
    def test_lance_a_slides_north(self) -> None:
        """lance_001: Lance A slides north."""
        board: BoardState = {
            "0,2": Piece(type="lance", color="white", variant="A")
        }
        result = validate_move(board, HexCoord(0, 2), HexCoord(0, -2), "white")
        assert result.legal is True
    
    def test_lance_a_cannot_move_ne(self) -> None:
        """lance_002: Lance A cannot move NE (not in its directions)."""
        board: BoardState = {
            "0,2": Piece(type="lance", color="white", variant="A")
        }
        result = validate_move(board, HexCoord(0, 2), HexCoord(2, 0), "white")
        assert result.legal is False


class TestChariotMoves:
    """Test chariot-specific move validation."""
    
    def test_chariot_slides_ne(self) -> None:
        """chariot_001: Chariot slides NE."""
        board: BoardState = {
            "0,0": Piece(type="chariot", color="white")
        }
        result = validate_move(board, HexCoord(0, 0), HexCoord(3, -3), "white")
        assert result.legal is True
    
    def test_chariot_cannot_move_north(self) -> None:
        """chariot_002: Chariot cannot move north (not diagonal)."""
        board: BoardState = {
            "0,0": Piece(type="chariot", color="white")
        }
        result = validate_move(board, HexCoord(0, 0), HexCoord(0, -2), "white")
        assert result.legal is False


class TestCheckValidation:
    """Test check-related move validation."""
    
    def test_king_cannot_move_into_check(self) -> None:
        """check_001: King cannot move into check."""
        board: BoardState = {
            "0,0": Piece(type="king", color="white"),
            "1,-4": Piece(type="queen", color="black")
        }
        result = validate_move(board, HexCoord(0, 0), HexCoord(1, 0), "white")
        assert result.legal is False
        assert result.reason == "movesIntoCheck"


class TestTurnValidation:
    """Test turn-related move validation."""
    
    def test_cannot_move_opponents_piece(self) -> None:
        """turn_001: Cannot move opponent's piece."""
        board: BoardState = {
            "0,-2": Piece(type="pawn", color="black")
        }
        result = validate_move(board, HexCoord(0, -2), HexCoord(0, -1), "white")
        assert result.legal is False
        assert result.reason == "notYourPiece"
    
    def test_cannot_move_from_empty_cell(self) -> None:
        """turn_002: Cannot move from empty cell."""
        board: BoardState = {
            "0,0": Piece(type="king", color="white")
        }
        result = validate_move(board, HexCoord(1, 0), HexCoord(2, 0), "white")
        assert result.legal is False
        assert result.reason == "noPieceAtSource"


# ============================================================================
# Test Coverage Report
# ============================================================================

class TestCoverageReport:
    """Report test coverage statistics."""
    
    @pytest.fixture(scope="class")
    def test_suite(self) -> dict[str, Any]:
        return load_test_suite()
    
    def test_coverage_report(self, test_suite: dict[str, Any]) -> None:
        """Print coverage statistics."""
        board_tests = [tc for tc in test_suite["testCases"] if tc["type"] == "boardValidation"]
        move_tests = [tc for tc in test_suite["testCases"] if tc["type"] == "moveValidation"]
        
        print("\n=== Spec Test Coverage Report (Python) ===")
        print(f"Board validation tests: {len(board_tests)}")
        print(f"Move validation tests: {len(move_tests)}")
        print(f"Total spec tests: {len(test_suite['testCases'])}")
        print("==========================================\n")
        
        assert len(test_suite["testCases"]) > 0
